### PR TITLE
Ensure editor selection buttons remain on screen when selection is near edge

### PR DESCRIPTION
**Pull request by peppy**
Addresses https://github.com/ppy/osu/discussions/23599.

Purely visual and if it regresses in the future, it's going to be in a completely unpredictable way, so opting to not add tests.

### CODE REVIEW

File: osu.Game/Screens/Edit/Compose/Components/SelectionBox.cs [hunk 1]
```diff
@@ -352,5 +362,29 @@ private void operationStarted()
...
+            float rightExcess = parentQuad.TopRight.X - buttonsQuad.TopRight.X;
+            float bottomExcess = thisQuad.BottomLeft.Y + buttonsQuad.Height - parentQuad.BottomLeft.Y;
+
+            if (leftExcess < 0)
+                buttons.X += ToLocalSpace(thisQuad.TopLeft - new Vector2(leftExcess)).X;
+            else if (rightExcess < 0)
+                buttons.X -= ToLocalSpace(thisQuad.TopLeft - new Vector2(rightExcess)).X;
+
+            if (bottomExcess > 0)
+                buttons.Y += ToLocalSpace(thisQuad.TopLeft - new Vector2(bottomExcess)).X;
```
Comment (bdach) [inline osu.Game/Screens/Edit/Compose/Components/SelectionBox.cs:387]: I kinda find this implementation to be a bit... off? It does keep the buttons on screen, but at the cost of sometimes ending up something like this:



which seems... subpar?

If I were to propose an alternative, it'd probably be something like this:

<details>
<summary>here's a patch</summary>

```diff
diff --git a/osu.Game/Screens/Edit/Compose/Components/SelectionBox.cs b/osu.Game/Screens/Edit/Compose/Components/SelectionBox.cs
index 677b0cd76e..05bf527b2d 100644
--- a/osu.Game/Screens/Edit/Compose/Components/SelectionBox.cs
+++ b/osu.Game/Screens/Edit/Compose/Components/SelectionBox.cs
@@ -4,12 +4,14 @@
 #nullable disable
 
 using System;
+using System.Linq;
 using osu.Framework.Allocation;
 using osu.Framework.Graphics;
 using osu.Framework.Graphics.Containers;
 using osu.Framework.Graphics.Shapes;
 using osu.Framework.Graphics.Sprites;
 using osu.Framework.Input.Events;
+using osu.Framework.Layout;
 using osu.Game.Graphics;
 using osu.Game.Graphics.Sprites;
 using osuTK;
@@ -247,8 +249,6 @@ private void recreate()
                     Height = 30,
                     Direction = FillDirection.Horizontal,
                     Margin = new MarginPadding(button_padding),
-                    Anchor = Anchor.BottomCentre,
-                    Origin = Anchor.TopCentre
                 }
             };
 
@@ -367,24 +367,50 @@ private void ensureButtonsOnScreen()
         {
             buttons.Position = Vector2.Zero;
 
-            var buttonsQuad = buttons.ScreenSpaceDrawQuad;
             var thisQuad = ScreenSpaceDrawQuad;
 
             // Shrink the parent quad to give a bit of padding so the buttons don't stick *right* on the border.
             // AABBFloat assumes no rotation. one would hope the whole editor is not being rotated.
             var parentQuad = Parent.ScreenSpaceDrawQuad.AABBFloat.Shrink(ToLocalSpace(thisQuad.TopLeft + new Vector2(button_padding * 2)));
 
-            float leftExcess = buttonsQuad.TopLeft.X - parentQuad.TopLeft.X;
-            float rightExcess = parentQuad.TopRight.X - buttonsQuad.TopRight.X;
-            float bottomExcess = thisQuad.BottomLeft.Y + buttonsQuad.Height - parentQuad.BottomLeft.Y;
+            float topExcess = thisQuad.TopLeft.Y - parentQuad.TopLeft.Y;
+            float bottomExcess = parentQuad.BottomLeft.Y - thisQuad.BottomLeft.Y;
+            float leftExcess = thisQuad.TopLeft.X - parentQuad.TopLeft.X;
+            float rightExcess = parentQuad.TopRight.X - thisQuad.TopRight.X;
 
-            if (leftExcess < 0)
-                buttons.X += ToLocalSpace(thisQuad.TopLeft - new Vector2(leftExcess)).X;
-            else if (rightExcess < 0)
-                buttons.X -= ToLocalSpace(thisQuad.TopLeft - new Vector2(rightExcess)).X;
+            if (leftExcess < 0 || rightExcess < 0)
+            {
+                buttons.Direction = FillDirection.Vertical;
+
+                if (leftExcess < 0)
+                {
+                    buttons.Anchor = Anchor.CentreRight;
+                    buttons.Origin = Anchor.CentreLeft;
+                }
+                else if (rightExcess < 0)
+                {
+                    buttons.Anchor = Anchor.CentreLeft;
+                    buttons.Origin = Anchor.CentreRight;
+                }
 
-            if (bottomExcess > 0)
-                buttons.Y += ToLocalSpace(thisQuad.TopLeft - new Vector2(bottomExcess)).X;
+                if (topExcess < 0) buttons.Y += ToLocalSpace(thisQuad.TopLeft - new Vector2(topExcess)).Y;
+                if (bottomExcess < 0) buttons.Y += ToLocalSpace(thisQuad.TopLeft + new Vector2(bottomExcess)).Y;
+
+                return;
+            }
+
+            buttons.Direction = FillDirection.Horizontal;
+
+            if (topExcess > bottomExcess)
+            {
+                buttons.Anchor = Anchor.TopCentre;
+                buttons.Origin = Anchor.BottomCentre;
+            }
+            else
+            {
+                buttons.Anchor = Anchor.BottomCentre;
+                buttons.Origin = Anchor.TopCentre;
+            }
         }
     }
 }
```

</details>

Preview of behaviour:

https://github.com/ppy/osu/assets/20418176/4afd64cb-1e0a-4ad8-b288-c8e8a5708dae

Comment (peppy) [reply osu.Game/Screens/Edit/Compose/Components/SelectionBox.cs:387]: Hmm, I'm not sure about having them change flow direction, as a user will probably get confused with the ordering..

Comment (bdach) [reply osu.Game/Screens/Edit/Compose/Components/SelectionBox.cs:387]: If changing direction is an eyesore then we can also do without.

<details>
<summary>patch 2</summary>

```diff
diff --git a/osu.Game/Screens/Edit/Compose/Components/SelectionBox.cs b/osu.Game/Screens/Edit/Compose/Components/SelectionBox.cs
index 677b0cd76e..fc57cbf17f 100644
--- a/osu.Game/Screens/Edit/Compose/Components/SelectionBox.cs
+++ b/osu.Game/Screens/Edit/Compose/Components/SelectionBox.cs
@@ -247,8 +247,6 @@ private void recreate()
                     Height = 30,
                     Direction = FillDirection.Horizontal,
                     Margin = new MarginPadding(button_padding),
-                    Anchor = Anchor.BottomCentre,
-                    Origin = Anchor.TopCentre
                 }
             };
 
@@ -367,24 +365,30 @@ private void ensureButtonsOnScreen()
         {
             buttons.Position = Vector2.Zero;
 
-            var buttonsQuad = buttons.ScreenSpaceDrawQuad;
             var thisQuad = ScreenSpaceDrawQuad;
 
             // Shrink the parent quad to give a bit of padding so the buttons don't stick *right* on the border.
             // AABBFloat assumes no rotation. one would hope the whole editor is not being rotated.
             var parentQuad = Parent.ScreenSpaceDrawQuad.AABBFloat.Shrink(ToLocalSpace(thisQuad.TopLeft + new Vector2(button_padding * 2)));
 
-            float leftExcess = buttonsQuad.TopLeft.X - parentQuad.TopLeft.X;
-            float rightExcess = parentQuad.TopRight.X - buttonsQuad.TopRight.X;
-            float bottomExcess = thisQuad.BottomLeft.Y + buttonsQuad.Height - parentQuad.BottomLeft.Y;
+            float topExcess = thisQuad.TopLeft.Y - parentQuad.TopLeft.Y;
+            float bottomExcess = parentQuad.BottomLeft.Y - thisQuad.BottomLeft.Y;
+            float leftExcess = thisQuad.TopLeft.X - parentQuad.TopLeft.X;
+            float rightExcess = parentQuad.TopRight.X - thisQuad.TopRight.X;
 
-            if (leftExcess < 0)
-                buttons.X += ToLocalSpace(thisQuad.TopLeft - new Vector2(leftExcess)).X;
-            else if (rightExcess < 0)
-                buttons.X -= ToLocalSpace(thisQuad.TopLeft - new Vector2(rightExcess)).X;
+            if (topExcess > bottomExcess)
+            {
+                buttons.Anchor = Anchor.TopCentre;
+                buttons.Origin = Anchor.BottomCentre;
+            }
+            else
+            {
+                buttons.Anchor = Anchor.BottomCentre;
+                buttons.Origin = Anchor.TopCentre;
+            }
 
-            if (bottomExcess > 0)
-                buttons.Y += ToLocalSpace(thisQuad.TopLeft - new Vector2(bottomExcess)).X;
+            if (leftExcess < 0) buttons.X += ToLocalSpace(thisQuad.TopLeft - new Vector2(leftExcess)).X;
+            if (rightExcess < 0) buttons.X += ToLocalSpace(thisQuad.TopLeft + new Vector2(rightExcess)).X;
         }
     }
 }
```

</details>

Preview:

https://github.com/ppy/osu/assets/20418176/22fd0cbc-3dde-484b-bba5-574f8efa9691

Kind of a halfway solution between yours and mine.

Comment (bdach) [reply osu.Game/Screens/Edit/Compose/Components/SelectionBox.cs]: I've pushed in the patch, plus a few extra changes in https://github.com/ppy/osu/pull/23649/commits/3ad5f8b9c9973788780e540cb98e5b260367df98. Extremely tall or extremely wide selections weren't still entirely behaving well so I applied slight adjustments there.

In particular, in very tall selections there is nowhere to put the buttons without switching orientation, so they go inside the select box as a last contingency, in a fixed position:



It's not worse than what you originally had so I'm pretty confident with going forward with this, but please double check for the sake of my peace of mind.

**Review by bdach — APPROVED** [review]

**Pull request merged**